hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java

```diff
@@ -0,0 +1,245 @@
+  // Map to store key-values in memory until it hits maxInMemorySizeInBytes
+  final private Map<T,R> inMemoryMap;
+  // Map to store key-valuemetadata important to find the values spilled to disk
+  final private DiskBasedMap<T,R> diskBasedMap;
+  // Schema used to de-serialize and readFromDisk the records written to disk
+  final private Schema schema;
+  // An estimate of the size of each payload written to this map
+  private volatile long estimatedPayloadSize = 0;
+  // TODO (na) : a dynamic sizing factor to ensure we have space for other objects in memory
+  final private Double sizingFactorForInMemoryMap = 0.8;
+
+  public ExternalSpillableMap(Long maxInMemorySizeInBytes, Schema schema,
+                              String payloadClazz, Optional<String> baseFilePath) throws IOException {
+    this.inMemoryMap = new HashMap<>();
+    this.diskBasedMap = new DiskBasedMap<>(schema, payloadClazz, baseFilePath);
+    this.maxInMemorySizeInBytes = (long) Math.floor(maxInMemorySizeInBytes*sizingFactorForInMemoryMap);
+    this.schema = schema;
+    this.currentInMemoryMapSize = 0L;
+  }
+
+  /**
+   * A custom iterator to wrap over iterating in-memory + disk spilled data
+   * @return
+   */
+  public Iterator<R> iterator() {
+    return new IteratorWrapper<>(inMemoryMap.values().iterator(), diskBasedMap.iterator());
+  }
+
+  /**
+   * Number of entries in DiskBasedMap
+   * @return
+   */
+  public int getDiskBasedMapNumEntries() {
+    return diskBasedMap.size();
+  }
+
+  /**
+   * Number of bytes spilled to disk
+   * @return
+   */
+  public long sizeOfFileOnDiskInBytes() {
+    return diskBasedMap.sizeOfFileOnDiskInBytes();
+  }
+
+  /**
+   * Number of entries in InMemoryMap
+   * @return
+   */
+  public int getInMemoryMapNumEntries() {
+    return inMemoryMap.size();
+  }
+
+  /**
+   * Approximate memory footprint of the in-memory map
+   * @return
+   */
+  public long getCurrentInMemoryMapSize() {
+    return currentInMemoryMapSize;
+  }
+
+  @Override
+  public int size() {
+    return inMemoryMap.size() + diskBasedMap.size();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return inMemoryMap.isEmpty() && diskBasedMap.isEmpty();
+  }
+
+  @Override
+  public boolean containsKey(Object key) {
+    return inMemoryMap.containsKey(key) || diskBasedMap.containsKey(key);
+  }
+
+  @Override
+  public boolean containsValue(Object value) {
+    return inMemoryMap.containsValue(value) || diskBasedMap.containsValue(value);
+  }
+
+  @Override
+  public R get(Object key) {
+    if(inMemoryMap.containsKey(key)) {
+      return inMemoryMap.get(key);
+    } else if(diskBasedMap.containsKey(key)) {
+      return diskBasedMap.get(key);
+    }
+    return null;
+  }
+
+  @Override
+  public R put(T key, R value) {
+    if (this.currentInMemoryMapSize < maxInMemorySizeInBytes || inMemoryMap.containsKey(key)) {
+      inMemoryMap.put(key, value);
+      // Naive approach for now
+      if(estimatedPayloadSize == 0) {
```

Comment:
@vinothchandar here

Reply:
again, we may benefit from generic size sampler class